unixfs/io/sizeadj.go

```diff
@@ -0,0 +1,109 @@
+package io
+
+import (
+	"errors"
+	"io"
+)
+
+func newSizeAdjReadSeekCloser(base ReadSeekCloser, size uint64) *sizeAdjReadSeekCloser {
+	r := &sizeAdjReadSeekCloser{base: base, size: int64(size)}
+	if r.size < 0 {
+		panic("size limited 2^63 − 1")
+	}
+	return r
+}
+
+type sizeAdjReadSeekCloser struct {
+	base   ReadSeekCloser
+	size   int64
+	offset int64
+}
+
+// Read implements the Read method as defined by io.Reader
+func (r *sizeAdjReadSeekCloser) Read(p []byte) (int, error) {
+	if r.offset >= r.size { // EOF
+		return 0, io.EOF
+	}
+	if int64(len(p)) > r.size-r.offset { // truncate
+		newsize := r.size - r.offset
+		p = p[:newsize]
+	}
+	n, err := r.base.Read(p)
+	if err == nil {
+		_, err = r.base.Read(nil)
```

Comment:
I'd just allow the short read and force the outer reader to loop. Unfortunately, a lot of readers don't like read attempts with empty buffers (and I'd rather not do two reads per outer read).

Reply:
I did, that initially and it broke some code I will look into it more closely.

Reply:
That's *not* good...